dask_cuda/explicit_comms/comms.py

```diff
@@ -0,0 +1,238 @@
+        self.client = client if client is not None else default_client()
+        self.sessionId = uuid.uuid4().bytes
+
+        # Get address of all workers (not Nanny addresses)
+        self.worker_addresses = list(self.client.run(lambda: 42).keys())
+
+        # Make all workers listen and get all listen addresses
+        self.worker_direct_addresses = []
+        for rank, address in enumerate(self.worker_addresses):
+            self.worker_direct_addresses.append(
+                self.submit(
+                    address,
+                    _create_listeners,
+                    len(self.worker_addresses),
+                    rank,
+                    wait=True,
+                )
+            )
+
+        # Each worker creates a UCX endpoint to all workers with greater rank
+        self.run(_create_endpoints, self.worker_direct_addresses)
+
+        # At this point all workers should have a rank and endpoints to
+        # all other workers thus we can now stop the listening.
+        self.run(_stop_ucp_listeners)
+
+    def submit(self, worker, coroutine, *args, wait=False):
+        """Run a coroutine on a single worker
+
+        Parameters
+        ----------
+        worker: str
+            Worker to run the `coroutine`
+        coroutine: coroutine
+            The function to run on the worker
+        *args:
+            Arguments for `coroutine`
+        wait: boolean, optional
+            If True, waits for the coroutine to finished before returning.
+        Returns
+        -------
+        ret: object or Future
+            If wait=True, the result of `coroutine`
+            If wait=False, Future that can be waited on later.
+        """
+        ret = self.client.submit(
+            _run_coroutine_on_worker,
+            self.sessionId,
+            coroutine,
+            args,
+            workers=[worker],
+            pure=False,
+        )
+        return ret.result() if wait else ret
+
+    def run(self, coroutine, *args, workers=None):
+        """Run a coroutine on workers
+
+        Parameters
+        ----------
+        coroutine: coroutine
+            The function to run on each worker
+        *args:
+            Arguments for `coroutine`
+        workers: list, optional
+            List of workers. Default is all workers
+        Returns
+        -------
+        ret: list
+            List of the output from each worker
+        """
+        if workers is None:
+            workers = self.worker_addresses
+        ret = []
+        for worker in workers:
+            ret.append(
+                self.client.submit(
+                    _run_coroutine_on_worker,
+                    self.sessionId,
+                    coroutine,
+                    args,
+                    workers=[worker],
+                    pure=False,
+                )
+            )
+        return self.client.gather(ret)
+
+    def dataframe_operation(self, coroutine, df_list, extra_args=tuple()):
```

Comment:
Perhaps there is a clean way to isolate all `dataframe`-specific code within a sub-director or dedicated file (like `explicit_comms/dataframe.py`)?  This way, we could also add `array`-specific code into something like `explicit_comms/array.py` in the future (if needed/desired).  Any thoughts on this?

Reply:
> This way, we could also add array-specific code into something like explicit_comms/array.py in the future (if needed/desired). Any thoughts on this?

Jumping in here.  I'd love to see us avoid this kind of thing in the future if we can.  I think that there are some long term drawbacks to going outside of mainline like this.  I understand that folks are pushing hard for GTC numbers, but eventually it would be good to figure out what was slowing down normal merges and fix the underlying problem rather than work around it.

Reply:
> but eventually it would be good to figure out what was slowing down normal merges and fix the underlying problem rather than work around it.

Agree mostly.  However, it does seem that the underlying problem (in many cases) is the scheduler's ability to handle the large/complex task graphs for complex/all-to-all shuffling.  The fact that dask relies on a centralized python-based scheduler makes it very difficult to match (or even approach) the performance of something like "explicit comms" for all-to-all communication.  

In order to enable users to mix workflows that include operations that dask already does well with those that are painfully slow (many-partition set_index/join/merge), I really do believe we need to enable users to inject isolated (optimized) tasks in place of much larger grpahs in specific/isolated cases where HPC-like performance is deemed necessary.  Ultimately, this is just a way to dramatically reduce the size of the task graph for specialized algorithms that dask traditionally struggles with.

Reply:
Ah!  So is it the scheduling overhead that's weighing this down?  How many tasks per second are we processing?  I wonder if there are ways to speed that up.  In principle I would expect Dask to be able to process a few thousand tasks per second.  I would be somewhat surprised if we were pushing up against that limit on a single DGX (but I'm often surprised these days :) )

Also, if anyone has a performance report for a dask-cudf merge I'd be very curious to see it.

Reply:
To clarify, scheduler *processing* doesn’t limit the performance of all-to-all shuffling.  Rather, the inefficiencies of things like unnecessary data movement seem to grow with the size of the graph. I think the problem has more to do with the fact that the scheduler needs to be flexible about *where* a large number of tasks will be executed. For optimal performance, we want to do things that require us to assume a (temporarily) static set of workers.  This assumption allows optimizations like aggregating data from multiple partitions when the data is persisted on the same worker (scaling algorithm complexity with the number of workers rather than the number of tasks - and increasing the typical message size).

  There is certainly room for improvement in the shuffle-based algorithms in mainline Dask.  My point is just that we would most likely need an unrealistically smart and performant scheduler to match *explicit comms*  performance when a large task graph can be replaced with a pre-defined execution pattern.

Note that I am still very open minded about this.

Reply:
If we're moving data around inappropriately then I think that we should figure out why that's happening and resolve that problem.  I somewhat doubt that that is happening though.

I think that there are certainly cases where we'll want to step outside of Dask for communication and computation, particularly in cases that require very low latency communication, or very fine grained control.

My intuition is that merges and sorts aren't in this class of problems.  I would expect about a 2x improvement by hand-crafting things at most.  I understand that you all are seeing significantly greater than 2x speed improvement, so I suspect that there is a bug somewhere.  I would like to try to isolate where that bug is, so that hopefully mainline dask merges with ucx-py can operate at near full speed (along with all other dask + cudf + ucx communications).  

I would love to see a performance report using dask-cudf + ucx on a DGX.  I would be very happy to look through this with folks if anyone is interested.  I suspect that we could find some surprising things that might lead us down some paths to explore.

Reply:
I think you are right @mrocklin - My earlier comments about the *need* for something like explicit comms for shuffling was mostly coming from my experience seeing a 40x speedup for `set_index`.  However, I just realized that this speedup is only ~2-3x when communication is performed with sockets/tcp.  Therefore, you are absolutely correct that the real problem is most-likely a bug in ucx for NVLink communication of small messages.  The explicit-comms rewrite avoids small messages by temporarily aggregating partitions for the shuffle.  Some of the remaining performance improvement I see for explicit comms is obtained by avoiding the need to assign a new “_partitions” column (and thus duplicate data).  I should be able to implement this last part in the mainline algorithm.  I will gather performance reports for merge/set_index on Monday and share :)   It would be great to get your thoughts.

Reply:
>   I will gather performance reports for merge/set_index on Monday and share :) It would be great to get your thoughts.

Woot!  I'm really looking forward to seeing these :)

Reply:
> the real problem is most-likely a bug in ucx for NVLink communication of small messages

This is not really a bug, but a known behavior when the CUDA IPC cache is disabled. However, enabling it today with explicit comms indeed triggers a [bug in UCX](https://github.com/openucx/ucx/issues/3192), and so far this explicit comms PR is the only reproducible I've seen, but we are still investigating it and perhaps we would be able to see better performance once it's fixed.

Reply:
Thanks for clarifying @pentschev!  I should have been more precise about the "bug."  Are you saying it is now **safe** to enable the CUDA IPC cache without explicit comms?  I was working under the (probably out-of-date) assumption that this would lead to OOM.

Reply:
> Are you saying it is now safe to enable the CUDA IPC cache without explicit comms?

Yes and no. One important thing to note is that CUDA IPC cache currently requires a memory pool (such as RMM's) to avoid OOM to control the entire application's memory management, as we are then safe on memory not being allocated/freed/reallocated, and we are currently working only with that scenario. However, enabling the pool for explicit comms has currently caused the bug I mentioned before (https://github.com/openucx/ucx/issues/3192), so probably it will crash currently, but not because it goes OOM. Finally, the case where no memory pool is used must still disable the cache to avoid OOMing.

Note that we've enabled the cache back in UCX-Py by default https://github.com/rapidsai/ucx-py/pull/397 again, as the main use cases we're focusing currently do use RMM pools. It's still possible to disable it by setting `UCX_CUDA_IPC_CACHE=n`.

Reply:
@mrocklin @rjzamora, I agree we should definitely try to move optimizations into the scheduler if possible, but I just want to mentioned that I have seen benchmarks that generates more than **200k tasks** and 1k partitions!

Reply:
i filed https://github.com/dask/distributed/issues/3413 in case folks want to dig deeper into questions around dask performance and joins

Reply:
Thanks @quasiben!  I will move future discussion about if/how something like explicit comms should be used over there :)